### ISSUE TITLE
Unknown target cpu : xtensa-esp32s3-espidf

### ISSUE THREAD

**Issue by lorenzo-pirchio** [briansmith/ring#2088]
Hi,
I'm working on a project on esp32s3, I need to use rustls but I have problems with the ring crate since xtensa is not supported .
I'm using the esp32-idf-template (https://github.com/esp-rs/esp-idf-template), and I changed the depedencies adding the ones I needed.
As suggested by @briansmith, I am using the latest commit on the main branch, checking if there is this change (https://github.com/briansmith/ring/commit/ed5b2a86cccfe30e670b33536fc7e494fb5e9e19), but It still doesn't work.
To be sure that I was using the latest commit, I downloaded the repo and added it as a local dependency.

*This is my cargo.toml*:

`[package]
name = "s3_rust"
version = "0.1.0"
edition = "2021"
resolver = "2"
rust-version = "1.71"

[[bin]]
name = "s3_rust"
harness = false # do not use the built in cargo test harness -> resolve rust-analyzer errors

[profile.release]
opt-level = "s"

[profile.dev]
debug = true    # Symbols are nice and they don't increase the size on Flash
opt-level = "z"

[features]
default = ["std", "embassy", "esp-idf-svc/native"]

pio = ["esp-idf-svc/pio"]
std = ["alloc", "esp-idf-svc/binstart", "esp-idf-svc/std"]
alloc = ["esp-idf-svc/alloc"]
nightly = ["esp-idf-svc/nightly"]
experimental = ["esp-idf-svc/experimental"]
embassy = ["esp-idf-svc/embassy-sync", "esp-idf-svc/critical-section", "esp-idf-svc/embassy-time-driver"]

[dependencies]
log = { version = "0.4", default-features = false }
esp-idf-svc = { version = "0.48", default-features = false }

[target.'cfg(target_os="espidf")'.dependencies]
ring = {path="C:/ring-main/ring-main", features =[ "less-safe-getrandom-espidf"]  }
rustls = { version = "0.23.7", default-features = false, features = ["ring", "logging", "std", "tls12"] }

[build-dependencies]
embuild = "0.31.3"
`

This is my config.toml:
`[build]
target = "xtensa-esp32s3-espidf"

[target.xtensa-esp32s3-espidf]
linker = "ldproxy"
runner = "espflash flash --monitor" # Select this runner for espflash v2.x.x
rustflags = [ "--cfg",  "espidf_time64"] # Extending time_t for ESP IDF 5: https://github.com/esp-rs/rust/issues/110

[unstable]
build-std = ["std", "panic_abort"]

[env]
MCU="esp32s3"
ESP_IDF_VERSION = "v5.1.3"
CRATE_CC_NO_DEFAULTS = "1"
`

This is what I get when I do cargo build:

warning: ring@0.17.8: cc1.exe: warning: C:\ring-main\ring-main\pregenerated: No such file or directory [-Wmissing-include-dirs]
warning: ring@0.17.8: In file included from C:\ring-main\ring-main\include/ring-core/base.h:74,
warning: ring@0.17.8:                  from C:\ring-main\ring-main\include/ring-core/mem.h:60,
warning: ring@0.17.8:                  from C:\ring-main\ring-main\crypto/curve25519/curve25519.c:22:
warning: ring@0.17.8: C:\ring-main\ring-main\include/ring-core/target.h:53:2: error: #error "Unknown target CPU"
warning: ring@0.17.8:    53 | #error "Unknown target CPU"
warning: ring@0.17.8:       |  ^~~~~
warning: ring@0.17.8: C:\ring-main\ring-main\include/ring-core/base.h:76:10: fatal error: ring_core_generated/prefix_symbols.h: No such file or directory
warning: ring@0.17.8:    76 | #include <ring_core_generated/prefix_symbols.h>
warning: ring@0.17.8:       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: ring@0.17.8: compilation terminated.

If any of you know what is the problem, please help me.
Thanks

**Comment by briansmith** [comment]
> This is what I get when I do cargo build:
> 
> warning: ring@0.17.8: cc1.exe: warning: C:\ring-main\ring-main\pregenerated: No such file or directory [-Wmissing-include-dirs]

It looks like you're building ring from source, but not from within a Git checkout. Thus, build.rs doesn't see `.git/` and so it is expecting some files in `pregenerated/`. But they are not there, because they only exist in the package that's uploaded to crates.io.

> `cargo build`

What is the exact command line of `cargo build`. As you can see in BUILDING.md, you need to set `TARGET_CC` and `TARGET_AR` to the correct C compiler.

**Comment by briansmith** [comment]
What does `xtensa-esp32s3-elf-gcc.exe --version` output?

**Comment by lorenzo-pirchio** [comment]
This is xtensa-esp32s3-elf-gcc.exe --version output:

xtensa-esp32s3-elf-gcc.exe (crosstool-NG esp-2021r2-patch5) 8.4.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

I added this to target.h:
#elif defined(__XTENSA__)
#define OPENSSL_32_BIT

Now, when I compile it, I don't get any errors or warnings, but I don't know actually if it will work fine.

**Comment by briansmith** [comment]
In target.h, you can see:
```
// Versions of GCC before 10.0 didn't define `__ILP32__` for all 32-bit targets.
#elif defined(__MIPSEL__) || defined(__MIPSEB__) || defined(__PPC__) || defined(__powerpc__) || defined(__csky__)
#define OPENSSL_32_BIT
```

Ideally, you would upgrade to a newer version of GCC. Then it should "just work." Otherwise, you can find out what symbol to add to the above `#elif` and add it for xtensa. Note `XTENSA` isn't the right thing; the symbol we're looking for will start with '__' like the others in the list.

**Comment by lorenzo-pirchio** [comment]
Yes, I wrote Xtensa with '__' as the beginning and end as you said, but in markdown double underscores are used to format text as bold, as you can see in the previous message.

**Comment by briansmith** [comment]
OK, it looks like somebody just needs to submit the PR.

**Comment by lorenzo-pirchio** [comment]
Do you want me to submit a PR with the lines I added for xtensa?

**Comment by briansmith** [comment]
> `#elif defined(__MIPSEL__) || defined(__MIPSEB__) || defined(__PPC__) || defined(__powerpc__) || defined(__csky__)`

^ Please extend that line.

Incidentally, is there no newer version of the xtensa SDK that lets you use GCC 10 or higher?

**Comment by lorenzo-pirchio** [comment]
I changed to GCC 12 because now I'm using esp-idf 5.1, which requires GCC 12.
But unfortunately I have problems with the linker and I have errors about ring. 
Here they are:
[Errors.txt](https://github.com/user-attachments/files/15899659/Errors.txt)
 I have been trying to find possible solutions but I'm stuck doint this for a week.
 If you have any suggestion, I'd be thankful to you.

**Comment by briansmith** [comment]
> I changed to GCC 12 because now I'm using esp-idf 5.1, which requires GCC 12.
> But unfortunately I have problems with the linker and I have errors about ring.
> Here they are:
> [Errors.txt](https://github.com/user-attachments/files/15899659/Errors.txt)
> I have been trying to find possible solutions but I'm stuck doint this for a week.
> If you have any suggestion, I'd be thankful to you.

https://www.esp32.com/viewtopic.php?t=1612 and other things seem to say that you may need `CFLAGS=-mlongcalls` when invoking cargo?

**Comment by juliankrieger** [comment]
@briansmith since I can't find a PR from @lorenzo-pirchio , I will create one myself.

Furthermore, I'm on GCC V 13 (via crosstools) and still encounter the same error.

**Comment by juliankrieger** [comment]
@briansmith I can confirm building works with 
- `espup install` (installs toolchains for esp32)
- the recent patch
- The following `.cargo/config.toml` section:

```
TARGET_CC="xtensa-esp32s3-elf-cc"
TARGET_AR="xtensa-esp32s3-elf-ar"
TARGET_CFLAGS="-mlongcalls"
```

**Comment by ealinye** [comment]
im also trying to build rustls with ring on esp32, here is a fragment of my `Cargo.toml` file
